interp/interp.go

```diff
@@ -439,25 +478,26 @@ func (interp *Interpreter) eval(src, name string, inc bool) (res reflect.Value,
 		return res, err
 	}
 
-	// Add main to list of functions to run, after all inits
-	if m := interp.main(); m != nil {
-		initNodes = append(initNodes, m)
-	}
-
 	if root.kind != fileStmt {
 		// REPL may skip package statement
 		setExec(root.start)
 	}
 	interp.mutex.Lock()
+	gs := interp.scopes[pkgName]
 	if interp.universe.sym[pkgName] == nil {
 		// Make the package visible under a path identical to its name
 		// TODO(mpl): srcPkg is supposed to be keyed by importPath. Verify it is necessary, and implement.
-		interp.srcPkg[pkgName] = interp.scopes[pkgName].sym
+		interp.srcPkg[pkgName] = gs.sym
 		interp.universe.sym[pkgName] = &symbol{kind: pkgSym, typ: &itype{cat: srcPkgT, path: pkgName}}
 		interp.pkgNames[pkgName] = pkgName
 	}
 	interp.mutex.Unlock()
 
+	// Add main to list of functions to run, after all inits
+	if m := gs.sym[mainID]; m != nil {
+		initNodes = append(initNodes, m.node)
+	}
```

Comment:
I don't think this is correct. I think we only want to add the main function to the scope that is called "main", right? But the way you have defined gs L486 does not guarantee that gs is main scope, since pkgName could be anything.
I think I had added that extra check at some point, because when inspecting the whole symbols map I was seeing extra "main" symbols that should not exist.

Reply:
if will replace L497 by:
```go
if m := gs.sym[mainID]; pkgName == mainID && m != nil {
```
to add the main symbol only if in a main packge

Reply:
as a tiny optimization, maybe it's faster to test pkgName == mainID before doing the rest of the test, but I'm not sure.

Reply:
This test is performed only once per call to `eval` or to `importSrc`. I would favor conciseness and readability here.